torch/distributed/algorithms/ddp_comm_hooks/post_localSGD_hook.py

```diff
@@ -0,0 +1,90 @@
+        "start_localSGD_iter",
+        "iter",
+    ]
+
+    def __init__(
+        self,
+        process_group,
+        subgroup,
+        start_localSGD_iter,
+    ):
+        logging.info(
+            "Local SGD will be started after {} iterations".format(start_localSGD_iter)
+        )
+
+        # The group used for all-reducing gradients globally.
+        self.process_group = process_group
+        # The group used for all-reducing gradients locally.
+        self.subgroup = subgroup
+        self.start_localSGD_iter = start_localSGD_iter
+        # Iteration/step in the training loop.
+        self.iter = 0
+
+    def maybe_increase_iter(self, bucket):
+        # Since bucket 0 is the last bucket to allreduce in an iteration.
+        # Only increase `iter` when bucket 0 is processed.
+        if bucket.is_the_last_bucket_to_allreduce():
+            self.iter += 1
+
+        if self.iter == self.start_localSGD_iter:
+            logging.info(
+                "Start to apply local SGD after {} iterations.".format(self.iter)
+            )
+
+
+def post_localSGD_hook(
+    state: PostLocalSGDState, bucket: dist.GradBucket
+) -> torch.futures.Future:
+    """
+    This DDP communication hook is used for running post-localSGD algorithm,
+    by combining with a model averaging component that runs after the optimizer step.
+
+    Args:
+        state (PostLocalSGDState): State information to run post-localSGD.
+            Users mainly need to tune ``start_localSGD_iter`` to determine when to start local SGD.
+        bucket (dist.GradBucket): Bucket that stores a 1D flattened gradient tensor that batches multiple per-variable tensors.
+            Note that since DDP comm hook only supports single process single device mode at this time,
+            only exactly one tensor is stored in this bucket.
+
+    Returns:
+        Future handler of the communication, which updates the gradients in place.
+
+    Example::
+        >>> state = PostLocalSGDState(process_group=process_group, subgroup=subgroup,
+                                  start_localSGD_iter=10)
+        >>> ddp_model.register_comm_hook(state, post_localSGD_hook)
+    """
+    global_group_to_use = (
+        state.process_group if state.process_group is not None else dist.group.WORLD
+    )
+    world_size = global_group_to_use.size()
+
+    # The input tensor is a flattened 1D tensor.
+    input_tensor = bucket.get_tensor()
+
+    # Run allreduce using `global_group_to_use` in the first `start_localSGD_iter` iterations.
+    if state.iter < state.start_localSGD_iter:
+        state.maybe_increase_iter(bucket)
+        return default._allreduce_fut(global_group_to_use, input_tensor)
+
+    # Run allreduce using `subgroup` after the first `start_localSGD_iter` iterations.
+    # From this moment, model averaging should run after the optimizer step,
+    # to globally allreduce all the parameters.
+    return default._allreduce_fut(state.subgroup, input_tensor)
```

Comment:
This hook won't actually result in local SGD being automatically enabled right, the user would still need to manually enable local SGD? It looks the hook controls within which subgroup to run the allreduce. Should we decouple it from local SGD concepts?

Reply:
> This hook won't actually result in local SGD being automatically enabled right, the user would still need to manually enable local SGD?

I will consider allreduce within subgroups as a kind of local update, which actually is a part of local SGD. Currently the user does need to enable the other model averaging module separately. Once we complete incorporating optimizer into DDP comm hook, then we can further append the model averaging component in the same hook.

> It looks the hook controls within which subgroup to run the allreduce. Should we decouple it from local SGD concepts?

The implementation here actually does not look as generic as it is. Once we don't do global allreduce in the comm hook, we are already doing local update as part of local SGD, and there must be a global model averaging at some time. Therefore, I don't think we need to decouple it from local SGD concept, and I cannot imagine how this hook can be used for other purposes.